### ISSUE TITLE
[css-lists-3] Inconsist `counter()` syntax between CSS 2 and 3

### ISSUE THREAD

**Issue by xiaochengh** [w3c/csswg-drafts#5795]
Should `content: counter(c, none)` be a parse-time failure, or a valid value that generates empty text?

In CSS 2.1:
- `counter() = counter(name, <list-style-type>?)` [1], where
- `<list-style-type> = disc | circle | square | decimal | decimal-leading-zero | lower-roman | upper-roman | lower-greek | lower-latin | upper-latin | armenian | georgian | lower-alpha | upper-alpha | none | inherit` [2]

So `content: counter(c, none)` is a valid declaration that doesn't generate any text. Chrome implements this behavior.

In CSS 3:
- `counter()  =  counter( <counter-name>, <counter-style>? )` [3], where
- `<counter-style> = <counter-style-name> | <symbols()>` [4], where
- `<counter-style-name>` is a `<custom-ident>` that is not an ASCII case-insensitive match for "none". [5]

So `content: counter(c, none)` is invalid at parse time. Firefox implements this behavior.

My proposal:  We should not allow `none`. Allowing `none` in CSS 2.1 looks like a spec bug to me, since it provides little value and doesn't make too much sense. (And the same syntax also allows `inherit`, which makes no sense at all).

[1] https://www.w3.org/TR/CSS2/generate.html#counter-styles
[2] https://www.w3.org/TR/CSS2/generate.html#propdef-list-style-type
[3] https://drafts.csswg.org/css-lists-3/#counter-functions
[4] https://drafts.csswg.org/css-counter-styles-3/#typedef-counter-style
[5] https://drafts.csswg.org/css-counter-styles-3/#typedef-counter-style-name

**Comment by SebastianZ** [comment]
It was already agreed on to make `none` an invalid value in #4163.

As the definition in CSS Lists 3 supercedes the one in CSS 2 (and CSS 2.2) and already excludes `none` as well as all CSS-wide keywords, I think nothing needs to be done here.

_If_ something needs to be corrected, it is [CSS 2.2 that still says that `counter(x, none)` yields an empty string](https://drafts.csswg.org/css2/#counter) instead of making it invalid.

Sebastian

**Comment by Loirooriol** [comment]
So what's strange here is that #4163 claimed that a `none` counter style was invalid in CSS2, and wanted to align css-lists-3 with that. But actually `none` was valid in CSS2 according to https://www.w3.org/TR/CSS2/syndata.html#counter ?

**Comment by fantasai** [comment]
@Loirooriol https://www.w3.org/TR/CSS2/generate.html#content

> `<counter>`
>    Counters may be specified with two different functions: 'counter()' or 'counters()'.  [...] **The name must not be 'none', 'inherit' or 'initial'. Such a name causes the declaration to be ignored.** 

Generally-speaking, prose restrictions override grammar leniency.

Agree with @SebastianZ that the example in syndata.html needs to be fixed, though. :/ That's definitely self-inconsistent.

**Comment by Loirooriol** [comment]
@fantasai Precisely, it says "The ***name*** must not be 'none'", given `counter(name, style)`. Here we are talking about the style, not the name.

**Comment by Loirooriol** [comment]
Also note https://www.w3.org/TR/CSS2/generate.html#counter-styles says

> All the styles are allowed, including 'disc', 'circle', 'square', **and 'none'**.
> ```css
> P:before         { content: counter(p, none) }
> ```

**Comment by SebastianZ** [comment]
@Loirooriol is right. The initial claim in #4163 that `counter(name, none)` and `counters(name, "string", none)` are invalid was wrong.

Though I agree with @xiaochengh that neither `<counter-name>` nor `<counter-style-name>` should allow `none`, i.e. align CSS 2.2 with CSS Counter Styles 3 for the same reasons as @fantasai outlined in #4163.

Sebastian

**Comment by Loirooriol** [comment]
The current situation is that WebKit and Gecko aligned made `counter(name, none)` as per the 2nd part of the previous resolution and css-lists, while Blink still accepts that as per the 1st part of the resolution and css2.

```html
<!DOCTYPE html>
<style>
div::after {
  content: "counter(name, none) is NOT valid";
  content: "counter(name, none) is valid" counter(name, none);
}
</style>
<div></div>
```

Agenda+ to confirm what to do, because the previous resolution is self-contradictory.
